### PR TITLE
Added Matching Engine + "Vertex Text Embedding API" notebook

### PR DESCRIPTION
**Pull request by ivanmkc**
Added the "Vertex Text Embedding API" version of Matching Engine notebook.

Done: Fix links in header before submission.

### CODE REVIEW

File: notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb [hunk 1]
```diff
@@ -0,0 +1,2028 @@
+{
```
Comment (andrewferlitsch) [inline notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>I know we have code in other notebooks to setup a VPC network,</p><p> </p><p>https://github.com/GoogleCloudPlatform/vertex-ai-samples/blob/main/notebooks/community/ml_ops/stage5/get_started_with_vertex_private_endpoints.ipynb</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/1876/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (andrewferlitsch) [inline notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>Perform online query -&gt; Perform an online query</p><p> </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/1876/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (andrewferlitsch) [inline notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>I have not needed quotes when using the [tag] syntax</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/1876/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (andrewferlitsch) [inline notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>recommend a single cell for all installs.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/1876/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='8'/>

Comment (andrewferlitsch) [inline notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>? why</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/1876/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='9'/>

Comment (andrewferlitsch) [inline notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>fyi&gt; you submitted with output from cells</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/1876/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='14'/>

Comment (andrewferlitsch) [inline notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>drop "will"</p><p> </p><p>perhaps some explanation on how you prepare the data</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/1876/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='31'/>

Comment (andrewferlitsch) [inline notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>I would combine these 2 code cells</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/1876/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='33'/>

Comment (andrewferlitsch) [inline notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>Add text cell to explain why you Extract the question ids and question text</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/1876/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='35'/>

Comment (andrewferlitsch) [inline notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <ul><li>Recommendations and -Clustering looks like there maybe a formatting issue.</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/1876/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='36'/>

Comment (andrewferlitsch) [inline notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #8.    model = TextEmbeddingModel.from_pretrained("textembedding-gecko-001")

Should this be inside the function?
</pre><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/1876/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='38'/>

Comment (andrewferlitsch) [inline notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>Add text cell to explain what code is for</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/1876/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='39'/>

Comment (andrewferlitsch) [inline notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>DIMENSIONS -&gt; print(DIMENSIONS)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/1876/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='43'/>

Comment (andrewferlitsch) [inline notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>text cell to explain code</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/1876/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='44'/>

Comment (andrewferlitsch) [inline notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>a JSON string -&gt; an individual JSON object</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/1876/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='45'/>

Comment (andrewferlitsch) [inline notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>heading?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/1876/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='47'/>

**Comment by andrewferlitsch** [comment]
Did a first DPE review pass

File: notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb (hunk 1 above)
Comment (sarahcdugan) [reply notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>Agree, and maybe add an article -- "Use the Stack Overflow dataset..."</p>

File: notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb [hunk 2]
```diff
@@ -0,0 +1,1483 @@
...
```
Comment (sarahcdugan) [inline notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>Update the second sentence to something like  "These are uploaded to the Vertex AI Matching Engine service, which is a high-scale, low-latency solution, for finding similar vectors from a large corpus. Matching Engine is a fully managed offering, further reducing operational overhead."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/1876/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (sarahcdugan) [inline notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>nit, add a period. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/1876/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (sarahcdugan) [inline notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>nit - remove the period from a header</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/1876/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (sarahcdugan) [inline notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>nit - you can remove all the periods here in the bulleted list </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/1876/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (sarahcdugan) [inline notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>Change second sentence to something like "To learn more about supported regions, see the <a href="https://cloud.google.com/vertex-ai/docs/general/locations" target="_blank">Vertex AI Locations page</a>."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/1876/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='15'/>

Comment (sarahcdugan) [inline notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>nit, remove end period. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/1876/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='25'/>

Comment (sarahcdugan) [inline notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>Broken link at Input data format and structure - https://cloud.google.com/vertex-ai/docs/matching-engine/match-eng-setup/format-structure</p><p> </p><p>Nit, remove period from subheader</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/1876/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='45'/>

Comment (sarahcdugan) [inline notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>Broken link </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/1876/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='54'/>

Comment (sarahcdugan) [inline notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>Change to "After you've built" or change to "After you build" </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/1876/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='68'/>

Comment (sarahcdugan) [inline notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>not - add period</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/1876/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='71'/>

Comment (sarahcdugan) [inline notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>Change section to</p><p> </p><p>"When you productionize this code into a service, you need to convert the nearest ID's returned from Vertex AI Matching Engine into data ready to be used by downstream services.</p><p>In this case, you need to convert the ID's to titles. You can use Google Cloud's Memorystore to deploy a managed Redis instance to save the id-title key-value pairs.</p><p>To learn more, see <a href="https://cloud.google.com/memorystore/docs/redis/create-manage-instances?hl=en" target="_blank">Create and manage Redis instances</a>."</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/1876/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='73'/>

Comment (ivanmkc) [reply notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>Done</p>

File: notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb (hunk 1 above)
Comment (ivanmkc) [reply notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>The missing context is that the eng team wants to showcase the new "public endpoint" that doesn't require VPC setup. We will keep one notebook for VPC setup but the rest will be rewritten to omit that.</p>

Comment (ivanmkc) [reply notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>Done</p>

Comment (ivanmkc) [reply notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>Some IDE's do need it unfortunately.</p>

Comment (ivanmkc) [reply notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>I forget which one. Maybe VSCode, but I don't exactly remember.</p>

Comment (ivanmkc) [reply notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>Forgot to remove it. </p>

File: notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb (hunk 2 above)
Comment (ivanmkc) [reply notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>Hm, all the notebooks that use the template will have this. Removed it in this case.</p>

Comment (ivanmkc) [reply notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>Done, but this is also in the template. Can you modify notebook_template.ipynb to meet your criteria so all future notebooks have this fix?</p>

File: notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb (hunk 1 above)
Comment (ivanmkc) [reply notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>Done</p>

Comment (ivanmkc) [reply notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>Done</p>

Comment (ivanmkc) [reply notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>Actually we can remove this cell.</p>

Comment (ivanmkc) [reply notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>This is a NBReview bug.</p>

Comment (ivanmkc) [reply notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>No, as the same model is used each time. Including it in the function will just add latency.</p>

Comment (ivanmkc) [reply notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>Done.</p>

Comment (ivanmkc) [reply notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>done</p>

Comment (ivanmkc) [reply notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>Done</p>

Comment (ivanmkc) [reply notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>Done</p>

File: notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb (hunk 2 above)
Comment (ivanmkc) [reply notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>Done</p>

File: notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb (hunk 1 above)
Comment (ivanmkc) [reply notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>No heading for this cell. Is one needed?</p>

File: notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb (hunk 2 above)
Comment (ivanmkc) [reply notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>Done</p>

Comment (ivanmkc) [reply notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>Done</p>

File: notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb (hunk 1 above)
Comment (andrewferlitsch) [reply notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>nope.</p>

**Comment by andrewferlitsch** [comment]
LGTM

File: notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb (hunk 2 above)
Comment (ivanmkc) [reply notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>done</p>

Comment (sarahcdugan) [reply notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>oooh great point. I'll take that bit of feedback to my team. Thanks for letting me know </p>

File: notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb [hunk 3]
```diff
@@ -0,0 +1,1438 @@
...
```
Comment (sarahcdugan) [inline notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>Extra period after 'overhead'</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/GoogleCloudPlatform/vertex-ai-samples/pull/1876/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

**Review by sarahcdugan — APPROVED** [review]

File: notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb (hunk 3 above)
Comment (ivanmkc) [reply notebooks/official/matching_engine/sdk_matching_engine_create_stack_overflow_embeddings_palm.ipynb:1]: <p>Oopsies.</p>

**Review by andrewferlitsch — APPROVED** [review]

**Pull request merged**